### PR TITLE
Add the URLFor optional method to the storagedriver api

### PR DESCRIPTION
**Pull request by AndreyKostov**
This implements part of proposal #49

### CODE REVIEW

File: storagedriver/s3/s3.go [hunk 1]
```diff
@@ -580,6 +580,16 @@ func (d *Driver) Delete(path string) error {
...
 }
 
+// URLFor returns a URL which may be used to retrieve the content stored at the given path.
+// May return an UnsupportedMethodErr in certain StorageDriver implementations.
+func (d *Driver) URLFor(path string) (string, error) {
+	if !storagedriver.PathRegexp.MatchString(path) {
+		return "", storagedriver.InvalidPathError{Path: path}
+	}
+
+	return d.Bucket.SignedURL(d.s3Path(path), time.Now().Add(24*time.Hour)), nil
```
Comment (BrianBland) [inline storagedriver/s3/s3.go]: For whatever reason the old registry created S3 URLs which were valid for 20 minutes ([reference](https://github.com/docker/docker-registry/blob/master/docker_registry/drivers/s3.py#L155)). I'm not sure what the recommended time frame should be, but I feel like 1 day is probably excessive.

**Comment by BrianBland** [comment]
Looks good overall, but we'll wait for some more review of #49.

File: storagedriver/s3/s3.go (hunk 1 above)
Comment (dmp42) [inline storagedriver/s3/s3.go]: Let the TTL be configurable.

Comment (BrianBland) [inline storagedriver/s3/s3.go]: We can add that as an argument to `URLFor()`, but certain drivers will ignore the TTL. As long as we define the semantic properly and stick with it then it's ok.

Comment (AndreyKostov) [inline storagedriver/s3/s3.go]: So, we are agreed on adding it as an argument to the URLFor?

Comment (BrianBland) [inline storagedriver/s3/s3.go]: I added that to the discussion over at #49. Leave it out for now, and we can add it when a decision is made.

File: storagedriver/filesystem/driver.go [hunk 2]
```diff
@@ -265,6 +265,12 @@ func (d *Driver) Delete(subPath string) error {
 	return err
 }
 
+// URLFor returns a URL which may be used to retrieve the content stored at the given path.
+// May return an UnsupportedMethodErr in certain StorageDriver implementations.
+func (d *Driver) URLFor(path string) (string, error) {
```
Comment (stevvooe) [inline storagedriver/filesystem/driver.go]: I am starting to think we need a default implementation of the storage driver, with method stubs and simple implementations for `URLFor`, `PutContent`, `GetContent` and `Move`. Implementors can embed and replace as the driver dictates.

File: storagedriver/s3/s3.go (hunk 1 above)
Comment (stevvooe) [inline storagedriver/s3/s3.go]: I've suggested the `URLFor` rpc call take an options argument on #46.

**Comment by stevvooe** [comment]
This is looking good. Is there some documentation that needs updating with this addition?

**Comment by BrianBland** [comment]
@stevvooe No, this should be self-contained. I suggest we merge this PR as-is and I'll tweak the method definition in my fix for #49.

I like the idea of a default driver though. Let's create an issue for that.

**Comment by stevvooe** [comment]
Ok, LGTM!

**Pull request merged**